core/src/main/java/jenkins/security/ResourceDomainRootAction.java

```diff
@@ -0,0 +1,278 @@
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import static java.time.Instant.*;
+import static java.time.temporal.ChronoUnit.MINUTES;
+
+/**
+ * Root action serving {@link DirectoryBrowserSupport} instances
+ * on random URLs to support resource URLs (second domain).
+ *
+ * @see ResourceDomainConfiguration
+ * @see ResourceDomainFilter
+ *
+ * @since TODO
+ */
+@Extension
+@Restricted(NoExternalUse.class)
+public class ResourceDomainRootAction implements UnprotectedRootAction {
+
+    private static final Logger LOGGER = Logger.getLogger(ResourceDomainRootAction.class.getName());
+
+    @CheckForNull
+    @Override
+    public String getIconFileName() {
+        return null;
+    }
+
+    @CheckForNull
+    @Override
+    public String getDisplayName() {
+        return null;
+    }
+
+    @CheckForNull
+    @Override
+    public String getUrlName() {
+        return "static-files";
+    }
+
+    public void doIndex(StaplerRequest req, StaplerResponse rsp) throws IOException {
+        if (!ResourceDomainConfiguration.isResourceRequest(req)) {
+            rsp.sendError(404, "Cannot handle requests to this URL unless on Jenkins resource URL.");
+        } else {
+            rsp.sendError(404, ResourceDomainFilter.ERROR_RESPONSE);
+        }
+    }
+
+    public Object getDynamic(String id, StaplerRequest req, StaplerResponse rsp) throws Exception {
+        if (!ResourceDomainConfiguration.isResourceRequest(req)) {
+            rsp.sendError(404, "Cannot handle requests to this URL unless on Jenkins resource URL.");
+            return null;
+        }
+
+        Token token = Token.decode(id);
+        if (token == null) {
+            rsp.sendError(404, ResourceDomainFilter.ERROR_RESPONSE);
+            return null;
+        }
+
+        String authenticationName = token.username;
+        String browserUrl = token.path;
+
+
+        Instant creationDate = ofEpochMilli(token.timestamp);
+        if (creationDate.plus(VALID_FOR_MINUTES, MINUTES).isAfter(now()) && creationDate.isBefore(now())) {
+            return new InternalResourceRequest(browserUrl, authenticationName);
+        }
+
+        // too old, so redirect to the real file first
+        return new Redirection(browserUrl);
+    }
+
+    private static class Redirection {
+        private final String url;
+
+        public Redirection(String url) {
+            this.url = url;
+        }
+
+        public void doDynamic(StaplerRequest req, StaplerResponse rsp) throws IOException {
+            String restOfPath = req.getRestOfPath();
+
+            String url = Jenkins.get().getRootUrl() + this.url + restOfPath;
+            rsp.sendRedirect(302, url);
+        }
+    }
+
+    public String getRedirectUrl(Token token, String restOfPath) {
+        String resourceRootUrl = getResourceRootUrl();
+        if (!resourceRootUrl.endsWith("/")) {
+            resourceRootUrl += "/";
```

Comment:
I think it would be better for `ResourceDomainConfiguration.checkUrl` to require the trailing slash.

Reply:
That would be fine too.